Source/Core/VideoCommon/ShaderCache.cpp

```diff
@@ -102,11 +102,12 @@ const AbstractPipeline* ShaderCache::GetPipelineForUid(const GXPipelineUid& uid)
   if (it != m_gx_pipeline_cache.end() && !it->second.second)
     return it->second.first.get();
 
+  const bool exists_in_cache = it != m_gx_pipeline_cache.end();
   std::unique_ptr<AbstractPipeline> pipeline;
   std::optional<AbstractPipelineConfig> pipeline_config = GetGXPipelineConfig(uid);
   if (pipeline_config)
     pipeline = g_renderer->CreatePipeline(*pipeline_config);
-  if (g_ActiveConfig.bShaderCache)
+  if (g_ActiveConfig.bShaderCache && !exists_in_cache)
     AppendGXPipelineUID(uid);
```

Comment:
I dislike the logic of when to call AppendGXPipelineUID. What about moving this to InsertGXPipeline?

Reply:
InsertGXPipeline just does the map insert, which requires a valid pipeline. We call this when background compiling the existing UIDs, so the entries loaded from the cache would get appended again if we did this in Insert as-is.

Reply:
oh, I see why you've done it in this way :/